src/async/base_core.cpp

```diff
@@ -3,96 +3,142 @@
-    Submit();
+void BaseCore::SetWait(State state) noexcept {
+  YACLIB_ASSERT(state == kWaitDrop || state == kWaitStop);
+  if (!SetCallback(const_cast<NopeCounter<IRef>&>(kEmptyRef), state)) {
+    DecRef();
   }
 }
 
-void BaseCore::SetCallbackInline(InlineCore& callback, bool async) noexcept {
-  YACLIB_DEBUG(_callback != nullptr, "callback not empty, so future already used");
-  _callback.Reset(NoRefTag{}, &callback);
-  const auto new_state =
-    State{static_cast<char>(static_cast<char>(State::HasCallbackInline) + static_cast<char>(async))};
-  const auto old_state = _state.exchange(new_state, std::memory_order_acq_rel);
-  if (old_state == State::HasResult) {  //  TODO(MBkkt) rel + fence(acquire) instead of acq_rel
-    static_cast<InlineCore&>(*_callback).CallInline(*this, new_state);
-  }
+bool BaseCore::ResetWait() noexcept {
+  std::uint64_t expected = _callback.load(std::memory_order_relaxed);
+  return expected != kResult && _callback.compare_exchange_strong(expected, kEmpty, std::memory_order_relaxed);
 }
 
-bool BaseCore::SetWait(IRef& callback) noexcept {
-  YACLIB_DEBUG(_callback != nullptr, "callback not empty, so future already used");
-  _callback.Reset(NoRefTag{}, &callback);
-  auto expected = State::Empty;
-  if (!_state.compare_exchange_strong(expected, State::HasWait, std::memory_order_acq_rel)) {
-    _callback.Reset(NoRefTag{}, nullptr);  // This is mean we have Result
-    return false;
-  }
-  return true;
+bool BaseCore::Empty() const noexcept {
+  auto callback = _callback.load(std::memory_order_acquire);
+  YACLIB_ASSERT(callback == kEmpty || callback == kResult);
+  return callback == kEmpty;
 }
 
-bool BaseCore::ResetWait() noexcept {
-  auto expected = State::HasWait;
-  if (_state.load(std::memory_order_acquire) == expected &&
-      _state.compare_exchange_strong(expected, State::Empty, std::memory_order_acq_rel)) {
-    YACLIB_DEBUG(_callback == nullptr, "callback empty, but should be our wait function");
-    _callback.Reset(NoRefTag{}, nullptr);  // This is mean we don't have executed callback
-    return true;
+bool BaseCore::Alive() const noexcept {
+  const State callback{kMask & _callback.load(std::memory_order_acquire)};
+  YACLIB_DEBUG(callback == kResult, "No needed to check Alive after store kResult or from Future side");
+  return callback != kWaitStop;
+}
+
+IExecutorPtr& BaseCore::GetExecutor() noexcept {
+  return _executor;
+}
+
+void BaseCore::SetExecutor(IExecutorPtr executor) noexcept {
+  _executor = std::move(executor);
+}
+
+BaseCore::BaseCore(State callback) noexcept : _callback{callback}, _caller{&kEmptyRef} {
+}
+
+void BaseCore::SetResult() noexcept {
+  auto* caller = _caller;
+  YACLIB_ASSERT(std::exchange(_caller, &kEmptyRef) != nullptr);
+  caller->DecRef();
+  // auto expected = _callback.load(std::memory_order_acquire);
+  // if (expected == kEmpty || (expected & kMask) == kWaitNope) {
```

Comment:
Hmm I think it can be checked like ((expected & kMask) & ~kWaitNope) == 0?

Do not sure about will it better ot not.
No we always have 1 exchange + 1 load + optional(check load) cas

If here will be cas will be 2 load + 2 optional(check load) cas, I don't think it's definitely better so needs benchmark

Reply:
No, have here cas instead of exchange definitely bad idea because we need cycle for timed wait case